camayoc/db_serializer.py

```diff
@@ -0,0 +1,172 @@
+            sources_destination.mkdir(parents=True, exist_ok=True)
+
+            destination = sources_destination / f"connection_{job_id}.json"
+            if not self.__check_destination(destination):
+                continue
+
+            scanjob = ScanJob(client=self._client, _id=job_id)
+            all_connectionjobs = self.__list_paged(scanjob.connection_results)
+            self.__save_json(all_connectionjobs, destination)
+
+    def _serialize_scanjobs(self):
+        scans_file = self._destination / DBSERIALIZER_SCANS_FILE_PATH
+        with scans_file.open() as fh:
+            all_scans = json.load(fh)
+
+        scan_ids = [scan_json.get("id") for scan_json in all_scans]
+
+        for scan_id in scan_ids:
+            scans_destination = self._destination / DBSERIALIZER_SCANJOBS_DIR_PATH
+            scans_destination.mkdir(parents=True, exist_ok=True)
+
+            destination = scans_destination / f"jobs_{scan_id}.json"
+            if not self.__check_destination(destination):
+                continue
+
+            scanjob = ScanJob(client=self._client, scan_id=scan_id)
+            all_scanjobs = self.__list_paged(scanjob.list)
+            self.__save_json(all_scanjobs, destination)
+
+    def _serialize_reports(self):  # noqa: C901
+        def gen_report_ids(all_scans):
+            for scan_json in all_scans:
+                for job in scan_json.get("jobs", []):
+                    if report_id := job.get("report_id"):
+                        yield report_id
+
+        def extractfiles(tar, destination, files):
+            for member in tar.getmembers():
+                member_name = Path(member.name).name
+                if member_name not in files:
+                    continue
+                file_destination = destination / member_name
+                if not self.__check_destination(file_destination):
+                    continue
+                if tar_fh := tar.extractfile(member):
+                    file_destination.write_bytes(tar_fh.read())
+
+        scans_file = self._destination / DBSERIALIZER_SCANS_FILE_PATH
+        with scans_file.open() as fh:
+            all_scans = json.load(fh)
+
+        for report_id in gen_report_ids(all_scans):
+            report_destination = self._destination / DBSERIALIZER_REPORTS_DIR_PATH / str(report_id)
+            report_destination.mkdir(parents=True, exist_ok=True)
+
+            report = Report(client=self._client, _id=report_id)
+            gzip_response = report.reports_gzip()
+            gzip_response.raise_for_status()
+
+            tar_bytes = io.BytesIO(gzip_response.content)
+            with tarfile.open(fileobj=tar_bytes, mode="r:gz") as tar:
+                extractfiles(tar, report_destination, ("details.json", "aggregate.json"))
+
+    def __list_paged(self, obj_fn):
+        all_objs = []
+        page = 1
+        while True:
+            payload = {"page": page}
+            http_response = obj_fn(params=payload)
```

Comment:
I ran the script and looks like this method should take API throttling in consideration.

```
request path : '/api/v1/scans/31/jobs/?page=1'
request body : None
request headers : {'User-Agent': 'python-requests/2.32.3', 'Accept-Encoding': 'gzip, deflate', 'Accept': '*/*', 'Connection': 'keep-alive', 'Authorization': 'Token 0a3c3e0f42d7c67d6b6ea9240f66134f07ab822c'}
response code : 429
json_error_message : {'detail': 'Request was throttled. Expected available in 1 second.'}
```

Reply:
Unfortunately, there are many places in Camayoc where API throttling might be a problem.

In my mental model, it's discovery-ci responsibility to prepare environment for Camayoc. Throttling for authenticated users are set to 500/seconds since discovery-ci MR #128.

Reply:
Agreed.